### PR TITLE
Feature/alex layout

### PR DESCRIPTION
**Pull request by Alexerson**
I've implemented a completely new layout. 
So far it won't work on mobile. I need to work on this a bit later, but you can try it on desktop and tell me what you think. 

In essence: the cards are a bit smaller and the discard pile is way less "in your face". Also the keepers are a bit more visible.

Anyway, let me know, I can keep going in this direction, or we can evaluate something else. 

Also in this PR: I implemented a "gameProgression" that seems to make sense and I found a couple bugs in the new actions.

### CODE REVIEW

File: modules/php/Cards/Actions/ActionMoveACreeper.php [hunk 1]
```diff
@@ -16,14 +16,18 @@ public function __construct($cardId, $uniqueId)
     );
   }
 
-  public $interactionNeeded = "keeperSelection";
+  public $interactionNeeded = "keeperSelectionOther";
```
Comment (tomlowi) [inline modules/php/Cards/Actions/ActionMoveACreeper.php:20]: why this change? the card description says "Choose any Creeper in front of any player" - so you can select one of your own of one of any opponent, right?

File: modules/php/Cards/Actions/ActionStealSomething.php [hunk 2]
```diff
@@ -16,7 +16,7 @@ public function __construct($cardId, $uniqueId)
     );
   }
 
-  public $interactionNeeded = "keeperSelection";
+  public $interactionNeeded = "keeperSelectionOther";
```
Comment (tomlowi) [inline modules/php/Cards/Actions/ActionStealSomething.php:19]: in this case you are right

File: fluxx.game.php [hunk 3]
```diff
@@ -272,9 +274,13 @@ protected function getAllDatas()
...
-    // with Fluxx, that's always something of a 50/50 chance ?
-    return 50;
+    // With Fluxx, there is no way to know when the game ends.
+    // This ensures that:
+    // - game progression is >50% after the 4th round
+    // - it tends to 100% but never goes above it
+
+    $turns = self::getStat("turns_number");
+    return round(100 - 100 / exp($turns / 5));
   }
```
Comment (tomlowi) [inline fluxx.game.php:284]: Nice, that looks smart enough indeed!

**Comment by tomlowi** [comment]
Great, I will have a look at the new layout proposal this evening here.

File: modules/php/Cards/Actions/ActionMoveACreeper.php (hunk 1 above)
Comment (Alexerson) [reply modules/php/Cards/Actions/ActionMoveACreeper.php:20]: Yes, you're right. To be honest I didn't pay too much attention to it. I simply noticed this wasn't matching the name of the function in JS and did a quick fix to be able to focus on the layout.

Comment (tomlowi) [reply modules/php/Cards/Actions/ActionMoveACreeper.php:20]: Probably a merge issue from the creepers branch and the changes in the action resolves.
Originally there was only the "keeperSelection", and we changed that because some actions allow own selection and some only on other players.

**Comment by tomlowi** [comment]
Allright, general ideas about the new layout:
overall, I think it looks very promising so definitely worth going further with this.
I really like the way all player's keepers  are around the central board (and how this nicely scales with nr of players), and the more limited goal and discard section are a great improvement.

But some issues and more thoughts:
- you can only get 4 keepers in a row for the players above and below the central board, and only 3 for the players left and right. Probably on smaller computer screens it would be even less for the left and right parts, this gets messy if only 1 or 2 keepers in a row (checked: on my laptop screen left and right player indeed have only 1 keeper per row).
I think it would be great if we were able to get this to 6 or 7 keepers so it becomes a rare situation to have 2 rows of keepers (because that kind of breaks the nice look).
Maybe we can combine this again with a slightly smaller version of the on-table keepers, and a little more overlap?
And then only have player keeper sections above and below the central board.
This breaks the feeling of a "circle of players", but would certainly benefit the keeper visibility.
Depending on screen size, with 6 players you could have 3 keeper sections above and 3 below, with 4 players 2-2, and with 3 players you could see all keepers in play above the main board?
And for smaller widths, only the My keeper section on top and all other players below?

- That would maybe also allow to have the Goal / Basic Rules / Discard blocks on 1 row in the central board, giving more space to the Other Rules stock?

- Small comment: the "My hand" text should be in the correct player color, like the keeper section title.
Probably even better to name that "My keepers" for the active player, instead of its own name.

- With the Double Agenda, if a 3rd Goal is played this looks bad, would be nice if that could also fit (or maybe we could "hide" the new Goal until one is discarded, because you are not allowed to discard the just played Goal anyway?

- Have you investigated if there is a way to change the stock overlap dynamically depending on the number of cards?
That would be a great way to adapt the hand stock overlap when the hand size becomes larger than 10 or 12 or ...
No idea if that would be possible, haven't looked at it yet myself.

- Also something I just noticed now regarding layout: in the overall player boards the "hand count" and "keeper count" would be cleaner if they are aligned in 1 row, certainly with more players

**Comment by Alexerson** [comment]
Hey, thanks for your comments.
I agree with most of them indeed. Happy you like the general direction, I'll try to work more on this later this week and will try to come up with solutions. 

One thing I might explore is to reduce even more the size of the cards once they are played on the table. Except from "action", the other cards all have a pictogram that explain what the goal/keeper/creeper/rule is about, so I think it would be acceptable. 

For the general layout: yes, I really like the "table-like" layout, but maybe splitting differently would actually make things easier. We will have to find a solution for the mobile anyway. 
Maybe changing the layout completely when there is an interaction needed? Like show all the keepers above the rules when you ened to pick a keeper, put the rules above the keepers to discard a rule, etc. It might be too much changes, so not sure it's really a good idea, but I'll explore.

I didn't find a way to have "dynamic" overlap, but it might be there but not documented because I'm under the impression that I already played a game in BGA where it was done. 
In any case, it's definitively possible to have an event listener on the stock object and do the automatic update ourselves.

The 2 parts that I didn't revisit for now are the "rules" and the "player score section". But yes, it should obviously be all in one row.
I'll add the tooltips as well during my next work sessions.

**Comment by tomlowi** [comment]
*"Maybe changing the layout completely when there is an interaction needed?
Like show all the keepers above the rules when you ened to pick a keeper
..."*

> I feel like this would probably require a lot of effort for something I
am not really convinced would even be an improvement.
It could even become more confusing if everything starts changing places
and shifting around for every type of action?

You are always welcome to experiment with it if you think you can make it
work without too much trouble, but personally I would vote to skip this for
now?

*"  I didn't find a way to have "dynamic" overlap, but it might be there
but not documented because I'm under the impression that I already played a
game in BGA where it was done. "*

I only know of some card games (Solo Whist for instance) that have a
user-preference option to overlap cards or not.
So each player you can choose to have their hand in a single row with
overlapped cards, or on multiple rows with fully visible cards.
By default that requires a full game reload when the user changes the
option. Might be good enough to start with...



Op wo 20 jan. 2021 om 23:29 schreef Alexerson <notifications@github.com>:

**Comment by tomlowi** [comment]
Abandoning this, useful ideas were already incorporated
